developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-file-reader.ts

```diff
@@ -11,53 +11,123 @@ import { CompilerCallbacks } from "@keymanapp/developer-utils";
 import { XMLParser } from 'fast-xml-parser';
 import { util } from '@keymanapp/common-types';
 import { ConverterMessages } from '../converter-messages.js';
+import { SchemaValidators } from '@keymanapp/common-types';
+import { KeylayoutXMLSourceFile } from '../../../common/web/utils/src/types/keylayout/keylayout-xml.js';
 import boxXmlArray = util.boxXmlArray;
 
 export class KeylayoutFileReader {
 
   constructor(private callbacks: CompilerCallbacks /*,private options: CompilerOptions*/) { };
 
+  /**
+   * @returns true if valid, false if invalid
+   */
+
+  public validate(source: KeylayoutXMLSourceFile): boolean {
+
+    if (!SchemaValidators.default.keylayout(source)) {
+      for (const err of (<any>SchemaValidators.default.keylayout).errors) {
+        this.callbacks.reportMessage(ConverterMessages.Error_SchemaValidationError({
+          instancePath: err.instancePath,
+          keyword: err.keyword,
+          message: err.message || 'Unknown AJV Error', // docs say 'message' is optional if 'messages:false' in options
+          params: Object.entries(err.params || {}).sort().map(([k, v]) => `${k}="${v}"`).join(' '),
+        }));
+      }
```

Comment:
See above- i think the minimal change here would be to use the existing type and remove `ConverterMessages.Error_SchemaValidationError` - fixing line numbers could be a future pr

```suggestion
        this.callbacks.reportMessage(DeveloperUtilsMessages.Error_SchemaValidationError({
          instancePath: err.instancePath,
          keyword: err.keyword,
          message: err.message || 'Unknown AJV Error', // docs say 'message' is optional if 'messages:false' in options
          params: Object.entries(err.params || {}).sort().map(([k, v]) => `${k}="${v}"`).join(' '),
        }));
      }
```